src/compiler/crystal/compiler.cr

```diff
@@ -346,13 +350,37 @@ module Crystal
         object_arg = Process.quote_windows(object_names)
         output_arg = Process.quote_windows("/Fe#{output_filename}")
 
+        cl = CL
         link_args = [] of String
+
+        # if the compiler and the target both have the `msvc` flag, we are not
+        # cross-compiling and therefore we should attempt detecting MSVC's
+        # standard paths
+        {% if flag?(:msvc) %}
+          if msvc_path = Crystal::System::VisualStudio.find_latest_msvc_path
```

Comment:
The not so nice thing is that now we're unconditionally invoking `vswhere.exe` on *every* invokation of Crystal.

And on my VM it takes almost a minute to finish, consistently.

<pre>
> .\crystal2.exe eval 'p Time.local'
{"finding latest msvc path": 2021-11-30 15:50:13.593808900 -08:00 Local}
{"finding vs installations": 2021-11-30 15:50:13.615752800 -08:00 Local}
{"finding vswhere": 2021-11-30 15:50:13.617481600 -08:00 Local}
<b>{"found vswhere": 2021-11-30 15:50:13.619154100 -08:00 Local}
{"called vswhere": 2021-11-30 15:50:14.535144200 -08:00 Local}</b>
{"found vs installations": 2021-11-30 15:50:14.537205000 -08:00 Local}
{"checking a version": 2021-11-30 15:50:14.539411300 -08:00 Local}
{"finding win sdk libpath": 2021-11-30 15:50:14.541482200 -08:00 Local}
{"found win sdk libpath": 2021-11-30 15:50:14.543453600 -08:00 Local}
LINK : C:\Users\User\AppData\Local\crystal\cache\crystal-run-eval.tmp.exe not found or not built by the last incremental link; performing full link
2021-11-30 15:50:14.857111400 -08:00 Local
</pre>

Very strangely, calling `vswhere` manually in the same terminal somehow doesn't take so long!

```
> Get-Date -Format HH:mm:ss.fff; & 'C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe' -requires Microsoft.VisualStudio.Component.VC.Tools.x86.x64 -format json -sort | out-null; Get-Date -Format HH:mm:ss.fff
15:54:21.255
15:54:21.318
```

Does anyone have any idea about the difference?

<details>

```diff
diff --git a/src/compiler/crystal/compiler.cr b/src/compiler/crystal/compiler.cr
index ebfa4a292..bbbf6998a 100644
--- a/src/compiler/crystal/compiler.cr
+++ b/src/compiler/crystal/compiler.cr
@@ -357,8 +357,11 @@ module Crystal
         # cross-compiling and therefore we should attempt detecting MSVC's
         # standard paths
         {% if flag?(:msvc) %}
+          p "finding latest msvc path": ::Time.local
           if msvc_path = Crystal::System::VisualStudio.find_latest_msvc_path
+            p "finding win sdk libpath": ::Time.local
             if win_sdk_libpath = Crystal::System::WindowsSDK.find_win10_sdk_libpath
+              p "found win sdk libpath": ::Time.local
               host_bits = {{ flag?(:bits64) ? "x64" : "x86" }}
               target_bits = program.has_flag?("bits64") ? "x64" : "x86"

diff --git a/src/crystal/system/win32/visual_studio.cr b/src/crystal/system/win32/visual_studio.cr
index f26254c69..773f78af9 100644
--- a/src/crystal/system/win32/visual_studio.cr
+++ b/src/crystal/system/win32/visual_studio.cr
@@ -16,8 +16,11 @@ module Crystal::System::VisualStudio
   def self.find_latest_msvc_path : ::Path?
     # ported from https://github.com/microsoft/vswhere/wiki/Find-VC
     # Copyright (C) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+    p "finding vs installations": ::Time.local
     if vs_installations = get_vs_installations
+      p "found vs installations": ::Time.local
       vs_installations.each do |installation|
+        p "checking a version": ::Time.local
         version_path = ::File.join(installation.directory, "VC", "Auxiliary", "Build", "Microsoft.VCToolsVersion.default.txt")
         next unless ::File.file?(version_path)

@@ -30,8 +33,11 @@ module Crystal::System::VisualStudio
   end

   private def self.get_vs_installations : Array(Installation)?
+    p "finding vswhere": ::Time.local
     if vswhere_path = find_vswhere
+      p "found vswhere": ::Time.local
       vc_install_json = `#{::Process.quote(vswhere_path)} -requires Microsoft.VisualStudio.Component.VC.Tools.x86.x64 -sort -format json`.chomp
+      p "called vswhere": ::Time.local
       return if !$?.success? || vc_install_json.empty?

       Array(Installation).from_json(vc_install_json)
```

Reply:
Both for this reason and because maybe it's nice anyway to use `cl.exe` if it's already in path, do you think we should change the behavior to be like that - don't try to locate anything if `cl.exe` is already available?

It's just a question, not a blocking comment. My approval still stands.

Reply:
Maybe, but it can fail if e.g. we use `link.exe` directly instead because that often refers to coreutils link from Git Bash (such as on CI), so `PATH` should have the lowest priority.

Reply:
@oprypin It's almost a second. Not minute. But yeah, that's still quite a bunch.

As an option, we could accept `CL` as an environment variable and if that is set, we don't need to look up the path.

Reply:
We might be able to detect the MSVC developer command prompt too by checking for the `VSCMD_VER` environment variable. In that case it is probably reasonable to use the defaults.

Reply:
Let's do that in a follow-up

Reply:
@straight-shoota Of course I meant a second. No idea how that slip happened.

Reply:
Time is relative, so... 🤷 🤣